UnifiProtect: Create NVR device before loading platforms to fix via_device references

RaHehl opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Create NVR device before loading platforms to fix via_device references

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #158189
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [ ] I understand the code I am submitting and can explain how it works.
- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

RaHehl committed:
Create NVR device before loading platforms to fix via_device references

Fixes #158189

RaHehl committed:
Merge branch 'dev' into fix/unifiprotect-via-device

MS2620 commented:
This PR correctly ensures the NVR device is created before platforms load, resolving the broken via_device references. The implementation is straightforward, follows existing patterns, and includes a relevant test, which is great for long-term stability.

A couple of small improvements would help, such as adding a brief comment explaining why the device must be created first, and considering whether an extra test is needed to cover reload/duplicate scenarios. Otherwise, this looks solid and is a valuable fix for the integration.

RaHehl committed:
Use snapshot for NVR device test

As suggested by @zweckj, use snapshot testing instead of
individual assertions for better maintainability.

RaHehl commented on tests/components/unifiprotect/test_init.py:
```diff
@@ -44,6 +44,33 @@ def mock_user_can_write_nvr(request: pytest.FixtureRequest, ufp: MockUFPFixture)
+    )
+    assert nvr_device is not None
+    assert nvr_device.manufacturer == "Ubiquiti"
+    assert nvr_device.name == nvr.display_name
```
Hey, thanks a lot for the suggestion! I wasn't familiar with snapshot testing in Python yet

RaHehl commented:
> This PR correctly ensures the NVR device is created before platforms load, resolving the broken via_device references. The implementation is straightforward, follows existing patterns, and includes a relevant test, which is great for long-term stability.
> 
> A couple of small improvements would help, such as adding a brief comment explaining why the device must be created first, and considering whether an extra test is needed to cover reload/duplicate scenarios. Otherwise, this looks solid and is a valuable fix for the integration.

There's already a comment on lines 163-164 that explains this: "This ensures via_device references work for all device entities"

No additional test is needed because device_registry.async_get_or_create() is idempotent by design - it automatically handles reload and duplicate scenarios by returning the existing device if it already exists, preventing any duplicates.

RaHehl committed:
Merge branch 'dev' into fix/unifiprotect-via-device
